src/Editor/ImGui/ImGuizmo.cpp

```diff
@@ -0,0 +1,1719 @@
+            m16[10] = (tmp[2] * src[12] + tmp[7] * src[13] + tmp[10] * src[15]) - (tmp[3] * src[12] + tmp[6] * src[13] + tmp[11] * src[15]);
+            m16[11] = (tmp[5] * src[12] + tmp[8] * src[13] + tmp[11] * src[14]) - (tmp[4] * src[12] + tmp[9] * src[13] + tmp[10] * src[14]);
+            m16[12] = (tmp[2] * src[10] + tmp[5] * src[11] + tmp[1] * src[9]) - (tmp[4] * src[11] + tmp[0] * src[9] + tmp[3] * src[10]);
+            m16[13] = (tmp[8] * src[11] + tmp[0] * src[8] + tmp[7] * src[10]) - (tmp[6] * src[10] + tmp[9] * src[11] + tmp[1] * src[8]);
+            m16[14] = (tmp[6] * src[9] + tmp[11] * src[11] + tmp[3] * src[8]) - (tmp[10] * src[11] + tmp[2] * src[8] + tmp[7] * src[9]);
+            m16[15] = (tmp[10] * src[10] + tmp[4] * src[8] + tmp[9] * src[9]) - (tmp[8] * src[9] + tmp[11] * src[10] + tmp[5] * src[8]);
+
+            // calculate determinant
+            det = src[0] * m16[0] + src[1] * m16[1] + src[2] * m16[2] + src[3] * m16[3];
+
+            // calculate matrix inverse
+            float invdet = 1 / det;
+            for (int j = 0; j < 16; ++j)
+            {
+                m16[j] *= invdet;
+            }
+        }
+
+        return det;
+    }
+
+    void matrix_t::RotationAxis(const vec_t & axis, float angle)
+    {
+        float length2 = axis.LengthSq();
+        if (length2 < FLT_EPSILON)
+        {
+            SetToIdentity();
+            return;
+        }
+
+        vec_t n = axis * (1.f / sqrtf(length2));
+        float s = sinf(angle);
+        float c = cosf(angle);
+        float k = 1.f - c;
+
+        float xx = n.x * n.x * k + c;
+        float yy = n.y * n.y * k + c;
+        float zz = n.z * n.z * k + c;
+        float xy = n.x * n.y * k;
+        float yz = n.y * n.z * k;
+        float zx = n.z * n.x * k;
+        float xs = n.x * s;
+        float ys = n.y * s;
+        float zs = n.z * s;
+
+        m[0][0] = xx;
+        m[0][1] = xy + zs;
+        m[0][2] = zx - ys;
+        m[0][3] = 0.f;
+        m[1][0] = xy - zs;
+        m[1][1] = yy;
+        m[1][2] = yz + xs;
+        m[1][3] = 0.f;
+        m[2][0] = zx + ys;
+        m[2][1] = yz - xs;
+        m[2][2] = zz;
+        m[2][3] = 0.f;
+        m[3][0] = 0.f;
+        m[3][1] = 0.f;
+        m[3][2] = 0.f;
+        m[3][3] = 1.f;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // 
+
+    enum MOVETYPE
+    {
+        NONE,
+        MOVE_X,
+        MOVE_Y,
+        MOVE_Z,
+        MOVE_XY,
+        MOVE_XZ,
+        MOVE_YZ,
+        MOVE_SCREEN,
+        ROTATE_X,
+        ROTATE_Y,
+        ROTATE_Z,
+        ROTATE_SCREEN,
+        SCALE_X,
+        SCALE_Y,
+        SCALE_Z,
+        SCALE_XYZ,
+        BOUNDS
+    };
+
+    struct Context
+    {
+        Context() : mbUsing(false), mbEnable(true), mbUsingBounds(false)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mDrawList' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mModelScaleOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mCameraEye' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mCameraRight' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mCameraDir' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mCameraUp' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mRayOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mRayVector' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mScreenFactor' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mRelativeOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mTranslationPlan' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mTranslationPlanOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mMatrixOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mRotationVectorSource' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mRotationAngle' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mRotationAngleOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mScale' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mScaleValueOrigin' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mSaveMousePosx' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBelowAxisLimit' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBelowPlaneLimit' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mAxisFactor' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBoundsPivot' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBoundsAnchor' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBoundsPlan' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBoundsLocalPivot' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBoundsBestAxis' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mBoundsAxis' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Member variable 'Context::mCurrentOperation' is not initialized in the constructor. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)